Date: Tue, 24 Jan 2023 20:12:36 +0000
Subject: Revert "xen/x2apic: enable x2apic mode when supported for HVM"

This reverts commit c8980fcb210851138cb34c9a8cb0cf0c09f07bf9.

This commit causes issues with hibernation on Xen guests on thaw,
resulting in all CPUs aside from 0 failing to come up.

While the commit is meant to be functionally equivalent, Xen guests would
previously trigger on
	xen_feature(XENFEAT_hvm_pirqs) && xen_have_vector_callback
and not enable x2apic support. Now that they do, there appear to be
issues with x2apic support and the downstream hibernation patches. While
that is investigated, revert this change.

Signed-off-by: Samuel Mendoza-Jonas <samjonas@amazon.com>
(cherry picked from commit c5ca0e9935870267029040b21a08908e2c75bc17)
Signed-off-by: Maximilian Heyne <mheyne@amazon.de>

diff --git a/arch/x86/include/asm/xen/hypervisor.h b/arch/x86/include/asm/xen/hypervisor.h
--- a/arch/x86/include/asm/xen/hypervisor.h
+++ b/arch/x86/include/asm/xen/hypervisor.h
@@ -46,6 +46,20 @@ static inline uint32_t xen_cpuid_base(void)
 	return cpuid_base_hypervisor(XEN_SIGNATURE, 2);
 }
 
+#ifdef CONFIG_XEN
+extern bool __init xen_hvm_need_lapic(void);
+
+static inline bool __init xen_x2apic_para_available(void)
+{
+	return xen_hvm_need_lapic();
+}
+#else
+static inline bool __init xen_x2apic_para_available(void)
+{
+	return (xen_cpuid_base() != 0);
+}
+#endif
+
 struct pci_dev;
 
 #ifdef CONFIG_XEN_PV_DOM0
diff --git a/arch/x86/xen/enlighten_hvm.c b/arch/x86/xen/enlighten_hvm.c
--- a/arch/x86/xen/enlighten_hvm.c
+++ b/arch/x86/xen/enlighten_hvm.c
@@ -260,9 +260,15 @@ static __init int xen_parse_no_vector_callback(char *arg)
 }
 early_param("xen_no_vector_callback", xen_parse_no_vector_callback);
 
-static __init bool xen_x2apic_available(void)
+bool __init xen_hvm_need_lapic(void)
 {
-	return x2apic_supported();
+	if (xen_pv_domain())
+		return false;
+	if (!xen_hvm_domain())
+		return false;
+	if (xen_feature(XENFEAT_hvm_pirqs) && xen_have_vector_callback)
+		return false;
+	return true;
 }
 
 static bool __init msi_ext_dest_id(void)
@@ -333,7 +339,7 @@ struct hypervisor_x86 x86_hyper_xen_hvm __initdata = {
 	.detect                 = xen_platform_hvm,
 	.type			= X86_HYPER_XEN_HVM,
 	.init.init_platform     = xen_hvm_guest_init,
-	.init.x2apic_available  = xen_x2apic_available,
+	.init.x2apic_available  = xen_x2apic_para_available,
 	.init.init_mem_mapping	= xen_hvm_init_mem_mapping,
 	.init.guest_late_init	= xen_hvm_guest_late_init,
 	.init.msi_ext_dest_id   = msi_ext_dest_id,
